### PR TITLE
hsa runtime

### PR DESCRIPTION
**Pull request by nimlgen**
HSA patches in a series of:
1) HSA runtime
2) Simple HSA graph (like HIP rn)
3) One big HSA graph with transfers embedded
...

### CODE REVIEW

File: tinygrad/device.py [hunk 1]
```diff
@@ -131,7 +131,7 @@ def copy(self, dest, src):
     if hasattr(dest.allocator.device, "track_cross_buffer") and hasattr(src.allocator, "track_cross_device"):
       dest.allocator.device.track_cross_buffer.append(src)
       src.allocator.track_cross_device.add(dest.allocator.device)
-    dest.allocator.transfer(dest._buf, src._buf, dest.nbytes)
+    dest.allocator.transfer(dest._buf, src._buf, dest.nbytes, **{'src_dev': src.allocator.device, 'dest_dev': dest.allocator.device})
```
Comment (geohot) [inline tinygrad/device.py:134]: src_dev= and dest_dev=

You don't need the dict

File: tinygrad/runtime/driver/hsa.py [hunk 2]
```diff
@@ -0,0 +1,143 @@
+import ctypes
+import tinygrad.runtime.autogen.hsa as hsa
+from tinygrad.helpers import init_c_var
+
+def check(status):
+  if status != 0: raise RuntimeError(f"HSA Error {status}")
```
Comment (chaosagent) [inline tinygrad/runtime/driver/hsa.py:6]: Is there a nice map from this to strings?

```
  File "/home/chaos/tinygrad/tinygrad/runtime/ops_hsa.py", line 65, in _alloc
    check(hsa.hsa_amd_memory_pool_allocate(self.device.gpu_mempool, size, 0, ctypes.byref(buf := ctypes.c_void_p())))
  File "/home/chaos/tinygrad/tinygrad/runtime/driver/hsa.py", line 6, in check
    if status != 0: raise RuntimeError(f"HSA Error {status}")
RuntimeError: HSA Error 4104
```
^ this seems like OOM error

**Comment by chaosagent** [comment]
Trying this with resnet, will update this post with observations

- hitting ctrlc too many times during a run frees the vram but pegs gpu at 100%, and i need to reset the gpus to run again

**Comment by geohot** [comment]
Testing llama

`JIT=2 HIP=1 python3 examples/llama.py  --temperature=0 --count=50 --prompt="Hello." --timing`

Load speed 5 GB/s
Step time 10 ms


vs

`JIT=2 HSA=1 python3 examples/llama.py  --temperature=0 --count=50 --prompt="Hello." --timing`

Load speed 2 GB/s
Step time 12 ms

Seems like there's also a big lag to init HSA. Can this be improved?

**Comment by nimlgen** [comment]
Do you experience this lag consistently? `Load speed 2 GB/s` seems to be that low because of this lag, it should be on par with HIP.

Yeah the python time is a bit slower with HSA, but overall llama should be faster with HSA:
```
enqueue in  12.37 ms
total  40.71 ms, 24.56 tok/sec
```
vs
```
enqueue in  10.98 ms
total  43.75 ms, 22.86 tok/sec
```

**Comment by geohot** [comment]
Yea, the lag seems to be consistent. You don't see it all the time?

**Comment by geohot** [comment]
Congrats!

**Pull request merged**